finetune/full.py

```diff
@@ -74,47 +77,72 @@ def setup(
+        IOArgs(data_dir, checkpoint_dir, out_dir),
+        TrainArgs(
+            save_interval,
+            log_interval,
+            global_batch_size,
+            micro_batch_size,
+            num_warmup_epochs,
+            epochs=num_epochs,
+            epoch_size=train_epoch_size,
+        ),
+        EvalArgs(eval_interval, eval_max_new_tokens, eval_iters),
+        OptimizationArgs(learning_rate),
+        DataArgs(max_seq_length),
+    )
+
 
-    fabric.print(hparams)
-    main(fabric, data_dir, checkpoint_dir, out_dir, resume)
+def main(
+    fabric: L.Fabric,
+    devices: int,
+    resume: Union[bool, Path],
+    config: Config,
+    io_args: IOArgs,
+    train_args: TrainArgs,
+    eval_args: EvalArgs,
+    optimization_args: OptimizationArgs,
+    data_args: DataArgs,
+) -> None:
+    steps_per_epoch = train_args.epoch_size // devices // train_args.batch_size(devices)
+    lr_max_steps = train_args.epochs * steps_per_epoch
+    lr_warmup_steps = train_args.lr_warmup_epochs * steps_per_epoch
 
+    check_valid_checkpoint_dir(io_args.checkpoint_dir)
 
-def main(fabric: L.Fabric, data_dir: Path, checkpoint_dir: Path, out_dir: Path, resume: Union[bool, Path]) -> None:
-    check_valid_checkpoint_dir(checkpoint_dir)
     fabric.seed_everything(1337)  # same seed for every process to init model (FSDP)
 
     if fabric.global_rank == 0:
-        os.makedirs(out_dir, exist_ok=True)
+        os.makedirs(io_args.out_dir, exist_ok=True)
 
-    train_data = torch.load(data_dir / "train.pt")
-    val_data = torch.load(data_dir / "test.pt")
+    train_data = torch.load(io_args.data_dir / "train.pt")
+    val_data = torch.load(io_args.data_dir / "test.pt")
 
-    config = Config.from_name(name=checkpoint_dir.name)
-    checkpoint_path = checkpoint_dir / "lit_model.pth"
+    checkpoint_path = io_args.checkpoint_dir / "lit_model.pth"
     fabric.print(f"Loading model {str(checkpoint_path)!r} with {config.__dict__}")
     with fabric.init_module(empty_init=(devices > 1)):
         model = GPT(config)
 
     fabric.print(f"Number of trainable parameters: {num_parameters(model, requires_grad=True):,}")
 
     model = fabric.setup(model)
-    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate, weight_decay=weight_decay)
+    optimizer = torch.optim.AdamW(
+        model.parameters(), lr=optimization_args.learning_rate, weight_decay=optimization_args.weight_decay
+    )
     optimizer = fabric.setup_optimizers(optimizer)
-    scheduler = get_lr_scheduler(optimizer, warmup_steps=warmup_steps, max_steps=max_steps)
-    state = {
-        "model": model,
-        "optimizer": optimizer,
-        "scheduler": scheduler,
-        "hparams": hparams,
-        "iter_num": 0,
-        "step_count": 0,
-    }
+    scheduler = get_lr_scheduler(optimizer, warmup_steps=lr_warmup_steps, max_steps=lr_max_steps)
+    state = {"model": model, "optimizer": optimizer, "scheduler": scheduler, "iter_num": 0, "step_count": 0}
```

Comment:
@awaelchli Do you want this state to be made into an actual class or should we keep it freeform for now?

Reply:
I would prefer freeform or AttributeDict. Fabric needs it in a dict anyway, and these aren't configurable by the CLI.